### PR TITLE
Not using try!

### PR DESCRIPTION
**Pull request by tapwork**
Use a proper `do catch` or use try? in combo with `guard` or `if let`
and display the error to get more context in failure case.

### CODE REVIEW

**Comment by loudmouth** [comment]
Hey @tapwork thanks for bringing this issue up and also for getting a fix together! 

The `try!` issue has been in the back of my mind since I picked up this project, but as the persistence library is simply conforming to a protocol that the main SDK defines, I was unsure about how to bubble up thrown errors back through the SDK and to the client program. The messages to the `PersistenceIntegration` are obviously sent in an async callback from the network and I'd prefer to not require that every network callback `throws` as it will add unnecessary boilerplate for these callbacks in 90% of cases (i.e. when the consumer is not using a persistence integration).

I think the `assertionFailure()` is a good solution, but I'm curious if all `fatalError()` should be swapped for the `assertionFailure()` calls so that production applications won't crash...I'm not sure if it's preferred to crash the application on failure to create/update a CoreData entity.

**Comment by loudmouth** [comment]
Just in case my previous comment was unclear, I just wanted to hear your opinion on `assertionFailure()` vs `fatalError()` before merging.

**Comment by tapwork** [comment]
hey @loudmouth 
sorry for my late answer.
I also thought about `fatalError` vs. `assertionFailure`.
It should this:  `fatalError` only if the program can't really go on and `assertionFailure` if the program still could go on. I will update my PR. give 1-2 hours

File: Sources/SynchronizationManager.swift [hunk 1]
```diff
@@ -135,8 +135,12 @@ public class SynchronizationManager: PersistenceIntegration {
         if let fetched = fetched?.first {
             persistable = fetched
         } else {
-            persistable = try! self.persistentStore.create(type: type)
-            persistable.id = asset.id
+            do {
+                persistable = try self.persistentStore.create(type: type)
+                persistable.id = asset.id
+            } catch let error {
+                fatalError("Could not create the Asset persistent store\n \(error)")
```
Comment (tapwork) [inline Sources/SynchronizationManager.swift]: I would leave `fatalError` here (so even crash in live world).  There is no guarantee that the program will work properly without the model. (which is expected to exist). The creation of the model should only fail in really rare cases. If it fails something else is really broken.
@loudmouth what do you think?

File: Sources/SynchronizationManager.swift [hunk 2]
```diff
@@ -164,12 +168,15 @@ public class SynchronizationManager: PersistenceIntegration {
...
         } else {
-            persistable = try! self.persistentStore.create(type: type)
-            persistable.id = entry.id
-            persistable.createdAt = entry.sys.createdAt
+            do {
+                persistable = try self.persistentStore.create(type: type)
+                persistable.id = entry.id
+                persistable.createdAt = entry.sys.createdAt
+            } catch let error {
+                fatalError("Could not create the Entry persistent store\n \(error)")
```
Comment (tapwork) [inline Sources/SynchronizationManager.swift]: Same here like in https://github.com/contentful/contentful-persistence.swift/pull/26/files#diff-6a7142dc6fe34b7d28c35b8871cad516R142

File: Sources/SynchronizationManager.swift [hunk 3]
```diff
@@ -200,7 +207,11 @@ public class SynchronizationManager: PersistenceIntegration {
     }
 
     public func save() {
-        try! persistentStore.save()
+        do {
+            try persistentStore.save()
+        } catch let error {
+            assertionFailure("Could not save the persistent store\n \(error)")
```
Comment (tapwork) [inline Sources/SynchronizationManager.swift]: The program can live without a save

File: Sources/SynchronizationManager.swift [hunk 4]
```diff
@@ -216,16 +227,19 @@ public class SynchronizationManager: PersistenceIntegration {
 
     // Returns regular (non-relationship) field to property mappings.
     internal func propertyMapping(for entryType: EntryPersistable.Type, and fields: [FieldName: Any]) -> [FieldName: String] {
+        // Filter out user-defined properties that represent relationships.
+        guard let persistentRelationshipPropertyNames = try? persistentStore.relationships(for: entryType) else {
+            assertionFailure("Could not filter out user-defined properties that represent relationships.")
```
Comment (tapwork) [inline Sources/SynchronizationManager.swift]: I am not 100% sure if this is really necessary to filter out the items in order to have a stable running app.
@loudmouth what do you think?

File: Sources/SynchronizationManager.swift [hunk 5]
```diff
@@ -308,8 +326,12 @@ public class SynchronizationManager: PersistenceIntegration {
 
     fileprivate func fetchSpace() -> SyncSpacePersistable {
         let createNewPersistentSpace: () -> (SyncSpacePersistable) = {
-            let spacePersistable: SyncSpacePersistable = try! self.persistentStore.create(type: self.persistenceModel.spaceType)
-            return spacePersistable
+            do {
+                let spacePersistable: SyncSpacePersistable = try self.persistentStore.create(type: self.persistenceModel.spaceType)
+                return spacePersistable
+            } catch let error {
+                fatalError("Could not create the Sync Space persistent store\n \(error)")
```
Comment (tapwork) [inline Sources/SynchronizationManager.swift]: I think the space is needed for the whole model, right @loudmouth ?

**Comment by loudmouth** [comment]
I think your approach is correct 👍 

Just wanted to think about it. I'm going to merge now. Thanks again for the contribution! I'll add you to the changelog.

**Comment by loudmouth** [comment]
@tapwork I will probably not be publishing a new version for the next few days, so you will want to point to `master` in the meantime with your dependency manager.

**Comment by tapwork** [comment]
thanks @loudmouth 
I will take master

**Pull request merged**